# Numpy Financial
*Outlines a sleek python package for conducting financial analysis*
[numpy-financial 1.0.0](https://www.example.com)

This is going to outline how we can do financial accounting with python. Python allows us to do a lot more than the traditional way of using excel. We can build this data back into our software applications and webpages using python. Python is the "backend" programming language for many major companies / websites.

![](https://images.squarespace-cdn.com/content/v1/55aed2d3e4b02de83b586b0a/1514837863854-T3R3V8JPLGMF5OTFL38M/ke17ZwdGBToddI8pDm48kLR4HjMe16aCHIVcvgOZ31EUqsxRUqqbr1mOJYKfIPR7LoDQ9mXPOjoJoqy81S2I8N_N4V1vUb5AoIIIbLZhVYxCRW4BPu10St3TBAUQYVKcMZfVDM3pZOlP1hLv09trMZvJLJu3bjM6essiIMkL10uBHc6mzswZt8Q6_XEwuqt1/Accounting.jpg?format=2500w)

I am taking my notes on the package using a jupyter notebook, which gives us an environment to write markdown and python through blocks of code. It is a very flexible piece of technology that should be encouraged / used. 

In [2]:
# lets us run the notebook from within our django project
import django_initializer

## Future value
This first function will allow us to compute the future value for a given asset. This is normally used by investors and financial planners to estimate how much an investment made today will be worth in the future. Using and learning python is great because I am going to show you why investing your time in learning python is way more important than trying to hack all of your processes together in microsoft office products.

In [4]:
import numpy as np
import numpy_financial as npf

What is the future value after 10 years of saving 100 dollars now, with an additional monthly savings of $100? Assume the interest rate is 5% (annually) compounded monthly.

In [5]:
npf.fv(0.05/12, 10*12, -100, -100)

15692.928894335748

By convention, the negative sign represents cash flow out (i.e. money not available today). Thus, saving $100 a month at 5% annual interest leads to $15,692.93 available to spend in 10 years.

Let’s compare different interest rates from the example above. We can pass the function for future value an array that will calculate each value we give it and then return back an array with the answer for each value that was fed. 

In [7]:
"""
Here we are comparing what the future value is for 5%, 6%, or 7% interest rates on the same savings amount of $100 down, then an additional $100 per month invested.
"""
a = np.array((0.05, 0.06, 0.07))/12
npf.fv(a, 10*12, -100, -100)

array([15692.92889434, 16569.87435405, 17509.44688102])

Future value (FV) is the value of a current asset at some point in the future based on an assumed growth rate.
Investors are able to reasonably assume an investment's profit using the future value (FV) calculation.
Determining the future value (FV) of a market investment can be challenging because of the market's volatility.
There are two ways of calculating the future value (FV) of an asset: FV using simple interest and FV using compound interest.

​	  
FV=I×(1+(R×T))

WHERE:

I = Investment amount
R = Interest rate
T = Number of years

FV=I×(1+R^T)

WHERE:

I = Investment amount
R = Interest rate
T = Number of years
​	

## Interest of a given payment

This allows us to calculate amortization and generate a scheduler for paying off a specific loan.


In [8]:
# ------------------------------- #
# imports 
# ------------------------------- #

import numpy as np
import numpy_financial as npf

"""
question: what is the amortization schedule for a 1 year loan of $2500 at 8.24% interest per year compunded?
"""

# setting a principal payment
principal = 2500.00

# setting a period (+ 1 because in finance arrays index at 1)
per = np.arange(1*12) + 1

# ipmt and ppmt are types of payment, interest vs principal
ipmt = npf.ipmt(0.0824/12, per, 1*12, principal)
ppmt = npf.ppmt(0.0824/12, per, 1*12, principal)

# each element of the sum of ipmt and ppmt equal the total payment
pmt = npf.pmt(0.0824/12, 1*12, principal)
np.allclose(ipmt + ppmt, pmt)

True

In [9]:
# string formatting for displaying an amortization schedule
fmt = fmt = '{0:2d} {1:8.2f} {2:8.2f} {3:8.2f}'
for payment in per:
    index = payment - 1
    principal = principal + ppmt[index]
    print(fmt.format(payment, ppmt[index], ipmt[index], principal))

 1  -200.58   -17.17  2299.42
 2  -201.96   -15.79  2097.46
 3  -203.35   -14.40  1894.11
 4  -204.74   -13.01  1689.37
 5  -206.15   -11.60  1483.22
 6  -207.56   -10.18  1275.66
 7  -208.99    -8.76  1066.67
 8  -210.42    -7.32   856.25
 9  -211.87    -5.88   644.38
10  -213.32    -4.42   431.05
11  -214.79    -2.96   216.26
12  -216.26    -1.49    -0.00


In [10]:
# carve off just the interest payments
interestpd = np.sum(ipmt)
np.round(interestpd, 2)

-112.98

## Internal Rate of Return

The internal rate of return is a discount rate that makes the net present value of all cash flows equal to zero in a discounted cash flow analysis. We use it to estimate the profitibility of potential investments.

- Internal rate of return (IRR) is the annual rate of growth an investment is expected to generate.

- IRR is calculated using the same concept as NPV, except it sets the NPV equal to zero.

- IRR is ideal for analyzing capital budgeting projects to understand and compare potential rates of annual return over time.

The "average" periodically compounded rate of return that gives a net present value of 0.0

Suppose you invest 100 units and then make the following withdrawals at regular (fixed) intervals: 39, 59, 55, 20. Assuming the ending value is 0, one's 100 unit investment yields 173 units; however, due to the combination of compounding and the periodic withdrawals, the "average" rate of return is neither simply 0.73/4 nor (1.73)^0.25-1. 



In [7]:
# ------------------------------- #
# imports 
# ------------------------------- #

import numpy_financial as npf

"""
examples of how to calculate internal rate of return, the first element in the array is our investment ($100 = -100), the next are the payments. 

values --> array_like, shape(N,):
-- Input cash flows per time period. By convention, net “deposits” are negative and net “withdrawals” are positive. Thus, for example, at least the first element of values, which represents the initial investment, will typically be negative.

out --> float
Internal Rate of Return for periodic input values.
"""

print(round(npf.irr([-100, 39, 59, 55, 20]), 5))
print(round(npf.irr([-100, 0, 0, 74]), 5))
print(round(npf.irr([-100, 100, 0, -7]), 5))
print(round(npf.irr([-100, 100, 0, 7]), 5))
print(round(npf.irr([-5, 10.5, 1, -8, 1]), 5))

0.28095
-0.0955
-0.0833
0.06206
0.0886


In [11]:
"""
we can also calculate a modified internal rate of return with numpy. This will take in the array, just like the previous calculation for irr, but then will take a finance rate and a reinvest rate as parameters too.

finance rate: interest paid on the cash flows
reinvest rate: interest rate recieved on the cash flows upon reinvestment
"""

npf.mirr([-1000, 100, 50, 10, 15], .34, .21)

-0.2742274708267268

## Periodic Payments
We can calculate the number of periods it takes pay off a given loan as well. This is solved by this equation: 
- fv + pv*(1+rate)**nper + pmt*(1+rate*when)/rate*((1+rate)**nper-1) = 0

If rate = 0 then:
- fv + pv*(1+rate)**nper + pmt*(1+rate*when)/rate*((1+rate)**nper-1) = 0


In [12]:
# ------------------------------- #
# imports 
# ------------------------------- #

import numpy as np
import numpy_financial as npf

"""
question: if you only had $150/mo to pay towards the loan, how long would it take to pay-off a loan of $8,000 at 7% annual interest?
"""

print(np.round(npf.nper(0.07/12, -150, 8000), 5))


64.07335


In [14]:
"""
we can also thorw this an array to do a comparative analysis across different combinations of interest rates, payments, and amounts... send in a grid format
"""

npf.nper(*(np.ogrid[0.07/12: 0.08/12: 0.01/12, 
                    -150   :  -99   : 50     ,
                    8000   :  9001  : 1000]))

array([[[ 64.07334877,  74.06368256],
        [108.07548412, 127.99022654]],

       [[ 66.12443902,  76.87897353],
        [114.70165583, 137.90124779]]])

## Net Present Value
Net present value is the difference betwen the present value of cash inflows and the present value of cash outflows over a period of time. NPV is used in capital budgeting an dinvestment planning to analyze the profitability of a projected investment of project.... re: use watson studio for this...

- Net present value, or NPV, is used to calculate the total value today of a future stream of payments.

- If the NPV of a project or investment is positive, it means that the discounted present value of all future cash flows related to that project or investment will be positive, and therefore attractive.

- To calculate NPV you need to estimate future cash flows for each period and determine the correct discount rate. 

Warning from our sponsor (numpy): npv considers a series of cashflows starting in the present (t = 0). NPV can also be defined with a series of future cashflows, paid at the end, rather than the start, of each period. If future cashflows are used, the first cashflow values must be zeroed and added to the net present value of the future cashflows. This is demonstrated in the examples.

We are going to define two variables in order to calculate the net present value:

    (1) The discount rate.

    (2) The values of the time series of cash flows. The (fixed) time interval between cash flow “events” must be the same as that for which rate is given (i.e., if rate is
        per year, then precisely a year is understood to elapse between each cash flow event). By convention, investments or “deposits” are negative, income or “withdrawals
        are positive; values must begin with the initial investment, thus values (position 0) will typically be negative.

**WARNING**: npv considers a series of cashflows starting in the present (t = 0). NPV can also be defined with a series of future cashflows, paid at the end, rather than the start, of each period. If future cashflows are used, the first cashflow values (position 0) must be zeroed and added to the net present value of the future cashflows. This is demonstrated in the examples.

In [1]:
# ------------------------------- #
# imports 
# ------------------------------- #

import numpy as np
import numpy_financial as npf

"""
Consider a potential project with an initial investment of $40 000 and projected cashflows of $5 000, $8 000, $12 000 and $30 000 at the end of each period discounted at a rate of 8% per period. To find the project’s net present value:
"""

rate, cashflows = 0.08, [-40_000, 5_000, 8_000, 12_000, 30_000]
npf.npv(rate, cashflows).round(5)

3065.22267

In [2]:
"""
It may be preferable to split the projected cashflow into an initial investment and expected future cashflows. In this case, the value of the initial cashflow is zero and the initial investment is later added to the future cashflows net present value:
"""

initial_cashflow = cashflows[0]
cashflows[0] = 0
np.round(npf.npv(rate, cashflows) + initial_cashflow, 5)

3065.22267

Given:
a present value, pv (e.g., an amount borrowed)

a future value, fv (e.g., 0)

an interest rate compounded once per period, of which there are

nper total

and (optional) specification of whether payment is made at the beginning (when = {‘begin’, 1}) or the end (when = {‘end’, 0}) of each period

Return:
the (fixed) periodic payment.